File: probfilter-core/src/main/java/com/c6h5no2/probfilter/util/RandomIntGenerator.java
package com.c6h5no2.probfilter.util;

import java.io.Serializable;


/**
 * A mutable random number generator.
 */
public interface RandomIntGenerator extends Mutable, Serializable {
    /**
     * @return a non-negative random integer
     */
    int nextInt();

    /**
     * @param bound a positive integer
     * @return a non-negative random integer within {@code [0, bound)}
     */
    default int nextInt(int bound) {
        return nextInt() % bound;
    }

    /**
     * @return a <i>new instance</i> of {@link RandomIntGenerator} with the same state
     * @implNote This operation should be cheap.
     */
    RandomIntGenerator copy();
}

File: probfilter-core/src/main/scala/com/c6h5no2/probfilter/util/RandomIntGenerator.scala
package com.c6h5no2.probfilter.util


/**
 * An immutable random number generator.
 *
 * @todo Breaking Change! Previous Cuckoo Filters assume mutability of getInt (prev. nextInt).
 */
trait RandomIntGenerator extends Serializable {
  /**
   * @return a <i>new instance</i> of [[RandomIntGenerator]] with the next state
   * @note This operation should be cheap.
   */
  def nextState(): RandomIntGenerator

  /**
   * @return a non-negative random integer
   */
  def getInt: Int

  /**
   * @param bound a positive integer
   * @return a non-negative random integer within `[0, bound)`
   */
  def getInt(bound: Int): Int = {
    getInt % bound
  }
}
